core/src/main/java/tachyon/client/BlockOutStream.java

```diff
@@ -132,25 +139,14 @@ public boolean canWrite() {
 
   @Override
   public void close() throws IOException {
+    if (canWrite() && mBuffer.position() > 0) {
```

Comment:
The behavior if !canWrite should probably be to throw an exception rather than silently drop the last buffer's worth of data. I think we can just put this back inside the `if (!mClosed)` statement without the canWrite() check; the caller can still call cancel() with your other change in place.